tools/dataset.cpp

```diff
@@ -1 +1,199 @@
+        for (size_t value_idx = 0; value_idx != source_array->length(); ++value_idx) {
+            edge_t edge {
+                .source_id = source_array->Value(value_idx),
+                .target_id = target_array->Value(value_idx),
+                .id = edges ? edge_array->Value(value_idx) : ukv_key_t {},
+            };
+            array.push_back(edge);
+            if (array.size() == max_batch_size) {
+                upsert_one_batch(c, array);
+                array.clear();
+            }
+        }
+    }
+    if (array.size() != 0)
+        upsert_one_batch(c, array);
+}
+
+void import_parquet(ukv_graph_import_t& c, ukv_size_t max_batch_size) {
+
+    arrow::Status status;
+    arrow::MemoryPool* pool = arrow::default_memory_pool();
+
+    // Open File
+    auto maybe_input = arrow::io::ReadableFile::Open(c.paths_pattern);
+    return_if_error(maybe_input.ok(), c.error, 0, "Can't open file");
+    auto input = *maybe_input;
+
+    std::unique_ptr<parquet::arrow::FileReader> arrow_reader;
+    status = parquet::arrow::OpenFile(input, pool, &arrow_reader);
+    return_if_error(status.ok(), c.error, 0, "Can't open file");
+
+    // Read File into table
+    std::shared_ptr<arrow::Table> table;
+    status = arrow_reader->ReadTable(&table);
+    return_if_error(status.ok(), c.error, 0, "Can't read file");
+
+    fill_array(c, max_batch_size, table);
+}
+
+void import_csv(ukv_graph_import_t& c, ukv_size_t max_batch_size) {
+
+    arrow::io::IOContext io_context = arrow::io::default_io_context();
+    auto maybe_input = arrow::io::ReadableFile::Open(c.paths_pattern);
+    return_if_error(maybe_input.ok(), c.error, 0, "Can't open file");
+    std::shared_ptr<arrow::io::InputStream> input = *maybe_input;
+
+    auto read_options = arrow::csv::ReadOptions::Defaults();
+    auto parse_options = arrow::csv::ParseOptions::Defaults();
+    auto convert_options = arrow::csv::ConvertOptions::Defaults();
+
+    // Instantiate TableReader from input stream and options
+    auto maybe_reader = arrow::csv::TableReader::Make(io_context, input, read_options, parse_options, convert_options);
+    return_if_error(maybe_reader.ok(), c.error, 0, "Can't instatinate reader");
+    std::shared_ptr<arrow::csv::TableReader> reader = *maybe_reader;
+
+    // Read table from CSV file
+    auto maybe_table = reader->Read();
+    return_if_error(maybe_table.ok(), c.error, 0, "Can't read file");
+    std::shared_ptr<arrow::Table> table = *maybe_table;
+
+    fill_array(c, max_batch_size, table);
+}
+
+////////// Parsing with SIMDJSON //////////
+
+simdjson::ondemand::document& rewinded(simdjson::ondemand::document& doc) noexcept {
+    doc.rewind();
+    return doc;
+}
+
+simdjson::ondemand::object& rewinded(simdjson::ondemand::object& doc) noexcept {
+    doc.reset();
+    return doc;
+}
+
+void import_json(ukv_graph_import_t& c, ukv_size_t max_batch_size) {
+
+    std::vector<edge_t> array;
```

Comment:
cant we preallocate a max size vector?

Reply:
For small data, this may be overkill.

Reply:
These tools are only intended for large imports